### ISSUE TITLE
proposal: colocate tests with files being tested

### ISSUE THREAD

**Issue by dpaiton** [rctn/sparsecoding#83]
tl;dr -- move unit tests to be located next to the files they test to improve coverage discoverability.

I will do this if we have a consensus.

------------

My proposal is each python file has a `*_test.py` file associated with it that lives in the same folder. This file contains unit tests for objects implemented in the python file. We would still have some integration, example, or similar tests that only make sense inside a top-level "tests" folder.

### Arguments for
- test location and coverage discoverability is higher (imo)
  - Reduces the cognitive load of navigating between source files and a separate test suite
- tests can be run by a user after installing the package to ensure the install went well ([discussion](https://pythonchb.github.io/PythonTopics/where_to_put_tests.html#sub-package))
- easy to run sub-tests ([pytest goodpractices discussion](https://docs.pytest.org/en/stable/explanation/goodpractices.html#tests-as-part-of-application-code))
- easy maintenance & refactoring
  - if you move files around you don't have to repeat the process in a mirrored directory

### Arguments against
- tests will ship with production code, which inflates the package size. I think this is not a real issue for research code. And I think it's currently happening anyway.
- there is some work to reorganize the current repo structure

### Examples in the wild
Numpy does something similar, although they group tests into lowest level folders. For example, here's the [polynomial folder](https://github.com/numpy/numpy/tree/main/numpy/polynomial). 

This is not an industry standard in Python repos. Many big repos use top-level folders, like [Pytorch](https://github.com/pytorch/pytorch/tree/main/test), [matplotlib](https://github.com/matplotlib/matplotlib/tree/main/lib/matplotlib/tests), and [Pandas](https://github.com/pandas-dev/pandas/tree/main/pandas/tests). 

What I'm proposing is standard practice in lower-level languages like [Rust](https://doc.rust-lang.org/book/ch11-03-test-organization.html#:~:text=You%E2%80%99ll%20put%20unit%20tests%20in%20the%20src%20directory%20in%20each%20file%20with%20the%20code%20that%20they%E2%80%99re%20testing).

**Comment by belsten** [comment]
I like this idea. The tests being in a separate `test` directory makes it easy to forget them and unclear what the coverage is. A 1:1 `_test.py` for `.py` will fix this issue. And you're right, it won't change the package size since we're already including the tests in the package. And besides, they're pretty lightweight.

This gets a 👍 from me.

**Comment by HarrisonSantiago** [comment]
Sounds good, I can't imagine this repo ever gets to the point that package size becomes a concern.

**Comment by 9q9q** [comment]
i like this a lot. thanks dylan!

**Comment by dpaiton** [comment]
Addressed in https://github.com/rctn/sparsecoding/pull/89
